### PR TITLE
[FEATURE] Add ability to duplicate panels

### PR DESCRIPTION
**Pull request by juliepagano**
Signed-off-by: Julie Pagano <julie.pagano@chronosphere.io>

# Notes for reviewers

This PR currently inserts duplicates in a new row at the bottom of the dashboard (the same as "add panel" does). I'm going to work on a follow-up that is "smarter" about inserting the new panel next to the panel it is duplicated from. The algorithm for this to account for the various cases you can have with a complex grid is non-trivial, so I want to do that in a follow-up both to keep this PR simple and to deliver some incremental functionality.

# Checklist

- [X] Pull request has a descriptive title and context useful to a reviewer.
- [X] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [X] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).
- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.

# TODO
- [ ] Adjust placement of duplicate to match guidance from Mohit. See notes for reviewers for more details. 

# Screenshots

image

duplicate

### CODE REVIEW

File: dev/data/dashboard.json [hunk 1]
```diff
@@ -1604,8 +1604,8 @@
               {
                 "x": 0,
                 "y": 0,
-                "width": 12,
-                "height": 6,
+                "width": 6,
+                "height": 4,
```
Comment (juliepagano) [inline dev/data/dashboard.json:1608]: Modified to values different than the defaults to get more signal out of the e2e test that validates duplicates keep the dimensions of the original panel.

File: ui/dashboards/src/components/Panel/PanelHeader.tsx [hunk 2]
```diff
@@ -50,6 +52,22 @@ export function PanelHeader({ id, title, description, editHandlers, isHovered, s
...
+            aria-label={ARIA_LABEL_TEXT.duplicatePanel(title)}
+            size="small"
+            onClick={editHandlers.onDuplicatePanelClick}
+          >
+            <ContentCopy
+              fontSize="inherit"
+              sx={{
+                // Shrink this icon a little bit to look more consistent
+                // with the other icons in the header.
+                transform: 'scale(0.9)',
```
Comment (juliepagano) [inline ui/dashboards/src/components/Panel/PanelHeader.tsx:66]: TODO: sync with Mohit on addressing the issues with this icon.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  // Actions in here need to modify both Panels and Panel Groups state
+  DuplicatePanelSlice & PanelSlice & PanelGroupSlice,
+  Middleware,
+  [],
+  DuplicatePanelSlice
+> {
+  return (set) => ({
+    duplicatePanel(panelGroupItemId: PanelGroupItemId) {
+      set((state) => {
+        const panels = state.panels;
```
Comment (juliepagano) [inline ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:43]: TODO: take another pass at this code for validations before opening for review.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        });
+
+        const dupePanelKey = getValidPanelKey(panelKey, panels);
+
+        state.panels[dupePanelKey] = panelToDupe;
+
+        const layout: PanelGroupItemLayout = {
+          i: generateId().toString(),
+          x: 0,
+          y: getYForNewRow(group),
```
Comment (juliepagano) [inline ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:73]: TODO: pull out defaults used here and in `panel-editor-slice` into a shared util.

File: ui/dashboards/src/utils/functions.ts [hunk 5]
```diff
@@ -1,16 +0,0 @@
-// Copyright 2023 The Perses Authors
```
Comment (juliepagano) [inline ui/dashboards/src/utils/functions.ts:1]: Renamed this utility for clarity, but I added enough changes that git treated it like a delete.

File: ui/dashboards/src/utils/panelUtils.test.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      'popularKey-1': {
+        kind: 'Panel',
+        spec: {},
+      } as PanelDefinition,
+    };
+    expect(getValidPanelKey('popularKey', mockPanelDefs)).toBe('popularKey-2');
+    expect(getValidPanelKey('popularKey-1', mockPanelDefs)).toBe('popularKey-2');
+  });
+
+  test('does not duplicate counters when they are out of order', () => {
```
Comment (juliepagano) [inline ui/dashboards/src/utils/panelUtils.test.ts:47]: This is an edge case the helper function I refactored would not have handled and could lead to panels with duplicate identifiers.

File: ui/dashboards/src/utils/panelUtils.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+import { PanelDefinition } from '@perses-dev/core';
+import { PanelGroupDefinition } from '../context';
+
+export const removeWhiteSpacesAndSpecialCharacters = (str: string) => {
+  return str.replace(/\s+/g, '');
+};
+
+// Given a PanelGroup, will find the Y coordinate for adding a new row to the grid, taking into account the items present
+export function getYForNewRow(group: PanelGroupDefinition) {
```
Comment (juliepagano) [inline ui/dashboards/src/utils/panelUtils.ts:22]: Helper extracted as is from `panel-editor-slice`, so it could also be used in `duplicate-panel-slice`.

File: ui/dashboards/src/utils/panelUtils.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  let newRowY = 0;
+  for (const layout of group.itemLayouts) {
+    const itemMaxY = layout.y + layout.h;
+    if (itemMaxY > newRowY) {
+      newRowY = itemMaxY;
+    }
+  }
+  return newRowY;
+}
+
```
Comment (juliepagano) [inline ui/dashboards/src/utils/panelUtils.ts:32]: Helpers below extracted from `panel-editor-slice` and significantly modified to address the edge case I called out in the tests.

File: ui/e2e/src/pages/DashboardPage.ts [hunk 9]
```diff
@@ -214,35 +211,67 @@ export class DashboardPage {
...
+   */
+  getPanel(panelName: string): Panel {
     const container = this.panels.filter({
       has: this.page.getByRole('heading', { name: panelName }),
     });
     return new Panel(container);
   }
 
-  async editMarkdownPanel(panelName: string, { name, groupName }: EditMarkdownPanelConfig) {
-    const panel = this.getPanel(panelName);
```
Comment (juliepagano) [inline ui/e2e/src/pages/DashboardPage.ts:225]: I've been unhappy with this helper for a while, so I cleaned it up and replaced it with the more generic `editPanel` while I was touching it for the tests in this PR.

File: ui/e2e/src/tests/panels.spec.ts [hunk 10]
```diff
@@ -128,10 +134,67 @@ test.describe('Dashboard: Panels', () => {
       const smallPanelPercentSize = await panelGroup.getPanelPercentOfBounds(panel);
       expect(smallPanelPercentSize.width).toBeCloseTo(1, 1);
 
-      // Panel is ~50% of panel group on a larger screen.
+      // Panel returns to the original size, which should be ~25%.
       await page.setViewportSize(previousViewport);
       const largePanelPercentSize = await panelGroup.getPanelPercentOfBounds(panel);
-      expect(largePanelPercentSize.width).toBeCloseTo(0.5, 1);
+      expect(largePanelPercentSize).toEqual(originalPanelPercentSize);
+      expect(largePanelPercentSize.width).toBeCloseTo(0.25, 1);
```
Comment (juliepagano) [inline ui/e2e/src/tests/panels.spec.ts:141]: Adjusting this test to match the changes I made to the panel's size.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+          throw new Error(`Cannot find Panel with key '${panelKey}'`);
+        }
+
+        const matchingLayout = group.itemLayouts.find((itemLayout) => {
+          return itemLayout.i === panelGroupLayoutId;
+        });
+
+        const dupePanelKey = getValidPanelKey(panelKey, panels);
+
+        state.panels[dupePanelKey] = panelToDupe;
```
Comment (juliepagano) [inline ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:68]: Not sure this is safe -- dig into immer and cloning to figure that out.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+        // Find the panel to edit
+        const panelToDupe = panels[panelKey];
+        if (panelToDupe === undefined) {
+          throw new Error(`Cannot find Panel with key '${panelKey}'`);
+        }
+
+        const matchingLayout = group.itemLayouts.find((itemLayout) => {
+          return itemLayout.i === panelGroupLayoutId;
+        });
```
Comment (juliepagano) [inline ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:64]: Is there a better way to get this value? It seems like this could have perf implications on more complex dashboards.

File: ui/dashboards/src/components/Panel/PanelHeader.tsx [hunk 13]
```diff
@@ -50,6 +52,22 @@ export function PanelHeader({ id, title, description, editHandlers, isHovered, s
...
+            aria-label={ARIA_LABEL_TEXT.duplicatePanel(title)}
+            size="small"
+            onClick={editHandlers.onDuplicatePanelClick}
+          >
+            <ContentCopy
+              fontSize="inherit"
+              sx={{
+                // Shrink this icon a little bit to look more consistent
+                // with the other icons in the header.
+                transform: 'scale(0.925)',
```
Comment (juliepagano) [inline ui/dashboards/src/components/Panel/PanelHeader.tsx:66]: Reviewed this with Mohit and came up with this exact value based on that.

File: ui/dashboards/src/utils/panelUtils.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  }
+  return normalizedPanelKey;
+}
+
+type PanelKeyParts = {
+  name: string;
+  number?: number;
+};
+
+const removeWhiteSpacesAndSpecialCharacters = (str: string) => {
```
Comment (juliepagano) [inline ui/dashboards/src/utils/panelUtils.ts:50]: This helper was kept from when `functions.ts` got renamed, but otherwise hasn't changed. 

Should we rename this to just `removeWhiteSpaces`? Maybe I'm missing some historical context?

File: ui/dashboards/src/utils/panelUtils.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  return {
+    name: panelKey,
+  };
+}
+
+// Find all the unique panel keys and the largest number used for each.
+// ex: cpu, cpu-1, cpu-2 count as the same panel key since these panels have the same name
+function getUniquePanelKeys(panels: Record<string, PanelDefinition>): Record<string, number> {
```
Comment (juliepagano) [inline ui/dashboards/src/utils/panelUtils.ts:74]: This helper was modified significantly to account for the issue called out in the tests above. Previously, it would count the number of matching keys and set a new key based on that. This could run into issues as panels get added and removed. Now, the function looks for the largest number seen for a given key and uses that, which should be more durable.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+        state.panels[dupePanelKey] = panelToDupe;
+
+        const layout: PanelGroupItemLayout = {
+          i: generateId().toString(),
+          x: 0,
+          y: getYForNewRow(group),
+          w: matchingLayout.w,
+          h: matchingLayout.h,
+        };
```
Comment (juliepagano) [inline ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:81]: Defaulting to the same insertion behavior as the "add new panel" action for now. See note to reviewers for plans in a follow-up.

File: ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts (hunk 11 above)
Comment (eunicorn) [reply ui/dashboards/src/context/DashboardProvider/duplicate-panel-slice.ts:73]: yep this is safe to do because `state` here is basically a draft and immer will create new object based on this draft

File: ui/dashboards/src/utils/panelUtils.ts (hunk 14 above)
Comment (eunicorn) [reply ui/dashboards/src/utils/panelUtils.ts:50]: yep can rename it to `removeWhiteSpaces`

File: ui/dashboards/src/utils/panelUtils.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+  number?: number;
+};
+
+const removeWhiteSpacesAndSpecialCharacters = (str: string) => {
+  return str.replace(/\s+/g, '');
+};
+
+/**
+ * Breaks the specified panel key into the name and the optional `-number` used
+ * for deduping panels with the name name.
```
Comment (eunicorn) [inline ui/dashboards/src/utils/panelUtils.ts:56]:
```suggestion
 * for deduping panels with the same name.
```

**Review by eunicorn — APPROVED** [review]
🔥
